cpp/src/umap/knn_graph/algo.cuh

```diff
@@ -15,44 +15,119 @@
-
 namespace Algo {
 
-using namespace ML;
-
 /**
  * Initial implementation calls out to FAISS to do its work.
- * TODO: cuML kNN implementation should support FAISS' approx NN variants (e.g. IVFPQ GPU).
  */
 
-/**
- * void brute_force_knn(float **input, int *sizes, int n_params, IntType D,
-                     float *search_items, IntType n, int64_t *res_I, float *res_D,
-                     IntType k, cudaStream_t s)
- */
-template <typename T>
-void launcher(float *X, int x_n, float *X_query, int x_q_n, int d,
-              int64_t *knn_indices, T *knn_dists, int n_neighbors,
-              UMAPParams *params, std::shared_ptr<deviceAllocator> d_alloc,
+template <typename value_idx = int64_t, typename value_t = float,
+          typename umap_inputs>
+void launcher(const umap_inputs &inputsA, const umap_inputs &inputsB,
+              ML::knn_graph<value_idx, value_t> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream);
+
+// Instantiation for dense inputs, int64_t indices
+template <>
+void launcher(const ML::manifold_dense_inputs_t<float> &inputsA,
+              const ML::manifold_dense_inputs_t<float> &inputsB,
+              ML::knn_graph<int64_t, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
               cudaStream_t stream) {
   std::vector<float *> ptrs(1);
   std::vector<int> sizes(1);
-  ptrs[0] = X;
-  sizes[0] = x_n;
+  ptrs[0] = inputsA.X;
+  sizes[0] = inputsA.n;
+
+  MLCommon::Selection::brute_force_knn(
+    ptrs, sizes, inputsA.d, inputsB.X, inputsB.n, out.knn_indices,
+    out.knn_dists, n_neighbors, d_alloc, stream);
+}
 
-  MLCommon::Selection::brute_force_knn(ptrs, sizes, d, X_query, x_q_n,
-                                       knn_indices, knn_dists, n_neighbors,
-                                       d_alloc, stream);
+// Instantiation for dense inputs, int indices
+template <>
+void launcher(const ML::manifold_dense_inputs_t<float> &inputsA,
+              const ML::manifold_dense_inputs_t<float> &inputsB,
+              ML::knn_graph<int, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream) {
+  throw raft::exception("Dense KNN doesn't yet support 32-bit integer indices");
 }
+
+template <>
+void launcher(const ML::manifold_sparse_inputs_t<int, float> &inputsA,
+              const ML::manifold_sparse_inputs_t<int, float> &inputsB,
+              ML::knn_graph<int, float> &out, int n_neighbors,
+              const ML::UMAPParams *params,
+              std::shared_ptr<ML::deviceAllocator> d_alloc,
+              cudaStream_t stream) {
+  // TODO: Use handle as input here and remove manual creation
+  cusparseHandle_t cusparseHandle;
+  CUSPARSE_CHECK(cusparseCreate(&cusparseHandle));
```

Comment:
can we have raft::handle passed here instead of explicit cusparse handle creation?